FEATURE: Refactor publication and discarding endpoints

nezaniel opened the pull request:
This implements explicit endpoints for publishing and discarding sites or documents, adjusting Neos' behavior to the ESCR.

(see also: https://github.com/neos/neos-development-collection/pull/4943)

grebaldi committed:
TASK: Re-add flow annotations in backend service controller

grebaldi committed:
TASK: Add new publish/discard enpoints to Routes.yaml

grebaldi committed:
TASK: Add new publish/discard routes to backend connector

grebaldi committed:
TASK: Adjust Workspaces redux store to new publish/discard semantics

grebaldi committed:
!!!TASK: Refactor publication and discarding endpoints

grebaldi committed:
TASK: Wrap publishing endpoint invokations in command object

grebaldi committed:
TASK: Use @phpstan-param annotation for publish/discard actions

Otherwise Flow's reflection gets in the way of converting the request
body to array, because it doesn't know how to handle type
`array<string,string>`.

grebaldi committed:
TASK: Convert given `NodeAddress`es to `NodeAggregateId`s in all publish/discard actions

grebaldi committed:
TASK: Implement new publish workflow in redux-saga

grebaldi committed:
TASK: Adjust prop type in discard dialog

grebaldi committed:
BUGFIX: Fix reference to publish/discard-scope in `@connect` of DiscardDialog

grebaldi committed:
BUGFIX: Read `length` of publishable nodes to calculate `numberOfChangesToBeDiscarded` in DiscardDialog

grebaldi committed:
TASK: Turn PublishDiscardScope into enum and fix typo in references everywhere

grebaldi committed:
TASK: Introduce PublishDiscardMode

grebaldi committed:
TASK: Use site as removal attachment point if document is removed

grebaldi committed:
TASK: Implement new discard workflow in redux-saga

grebaldi committed:
TASK: Navigate to closest other document if current document has been discarded

grebaldi committed:
TASK: Fix linting errors

grebaldi commented:
* PHP linting check will be fixed once https://github.com/neos/neos-development-collection/pull/4943 has been merged
* JS unit tests will be green once https://github.com/neos/neos-ui/pull/3757 has been upmerged to `9.0`
* E2E tests are green, because of https://github.com/neos/neos-ui/pull/3752/commits/3a06560087b6c7736e5175dcbc4b6d312c653e8c - **We need to revert this commit before merging!**

grebaldi committed:
TASK: Temporarily add patch for neos/neos-development-collection#4943

This commit MUST be removed before merging PR #3752

mhsdesign commented on Classes/Domain/Model/Changes/Remove.php:
```diff
@@ -104,4 +104,17 @@ public function apply(): void
+    {
+        $subgraph = $this->contentRepositoryRegistry->subgraphForNode($this->subject);
+
+        if ($this->subject->nodeType->isOfType(NodeTypeNameFactory::NAME_DOCUMENT)) {
```
```suggestion
        if ($this->subject?->nodeType->isOfType(NodeTypeNameFactory::NAME_DOCUMENT)) {
```

otherwise one wouldnt be able to remove nodes without nodeType, or in Neos.Neos we use the `getNodeType` trait that provides a fallback. (NodeTypeWithFallbackProviderTrait or sth)

grebaldi replied:
Is `subject` nullable here or is it rather `subject->nodeType`? So:

```php
if ($this->subject->nodeType?->isOfType(NodeTypeNameFactory::NAME_DOCUMENT)) {
```

mhsdesign replied:
no sorry of course the latter :)

grebaldi replied:
Adjusted :)

mhsdesign commented on Classes/Domain/Model/Changes/Remove.php:
```diff
@@ -104,4 +104,17 @@ public function apply(): void
         }
     }
+
+    private function getRemovalAttachmentPoint(): ?NodeAggregateId
```
as pointed out here: https://github.com/neos/neos-development-collection/pull/4943#pullrequestreview-1949957107

this logic should imo reside in Neos.Neos so that there is an API way of deleting nodes so they will still be deletable via the workspace publisher.

grebaldi replied:
Indeed. Neos.Neos should provide a high level API for all change operations imho. But I would argue that this is out of scope for this PR.

We definitely need to straighten out the boundaries between Neos.Neos and Neos.Neos.Ui. I believe however that by now everything in Neos.Neos.Ui is treated as internal (mostly thanks to your effort :)), so it would be permissible for a deeper refactoring to take place post-release.

wdyt?

mhsdesign commented on Classes/ContentRepository/Service/WorkspaceService.php:
```diff
@@ -35,6 +34,11 @@
  */
 class WorkspaceService
 {
+    private const NODE_HAS_BEEN_CREATED = 0b0001;
+    private const NODE_HAS_BEEN_CHANGED = 0b0010;
+    private const NODE_HAS_BEEN_MOVED = 0b0100;
+    private const NODE_HAS_BEEN_DELETED = 0b1000;
```
that seems clever and cost effective but isnt this a bit hard to grasp as the information is interpreted in the js codbase?

grebaldi replied:
In the other PR #3759, I've already refactored the client side part of this: https://github.com/neos/neos-ui/pull/3759/commits/3ce8499159180c8a6ed3e3a39944fe6a5e908024

Bitmasks work in JS just as they do in PHP, so this can simply be mirrored.

mhsdesign commented:
i added my comments over here https://github.com/neos/neos-development-collection/pull/4943#pullrequestreview-1949957107

also we should probably keep in mind that at some point we want to publish the node removal more transparently see
https://github.com/neos/neos-ui/pull/2974

i guess the whole `removalAttachmentPoint` is hacky and we will need to change it to support this case ...

as noted above i think all the change projection logic should reside in Neos.Neos so we can refactor this anytime.
The change projection is market as experimental and unstable, so i would like to avoid it to leak into the NeosUi and to all places where people are interested in the current changes.

Maybe this is doable in the scope of this pr maybe not. Just an idea. Thanks for the refactoring so far.

grebaldi committed:
TASK: Update command and method names according to upstream changes

see:
https://github.com/neos/neos-development-collection/pull/4943/commits/2a86e89cd801e0a8745f68ddc29c60eb68ae5397

grebaldi committed:
TASK: Adjust publishing action names to match the Neos.Neos API

grebaldi committed:
TASK: Consider nullability of nodeType in Remove change object

Commit:
Adjust to new workspace model

Commit:
Introduce ChangeTargetWorkspace

grebaldi committed:
TASK: Add SyncWorkspace application command and adjust BackendServiceController accordingly

Commit:
Adjust BackendServiceController to WorkspaceRebaseFailed exception

Commit:
Revert "Adjust BackendServiceController to WorkspaceRebaseFailed exception"

This reverts commit 820942f4caab615523f0429c452117b242d022b8.

mhsdesign commented:
Thanks for all the adjustments and discussions so far. It seems the mutations are now handled by the Neos `Workspace` but we dont use the new `Workspace::countChangesInSite` api. Should we do this in a followup or later :D?

grebaldi commented:
@mhsdesign:
> Thanks for all the adjustments and discussions so far. It seems the mutations are now handled by the Neos `Workspace` but we dont use the new `Workspace::countChangesInSite` api. Should we do this in a followup or later :D?

Definitely later. Currently, we need the actual changes in the UI to calculate dirty-state in various places. The new API doesn't solve this for us yet, but that's an implementation detail and can be adjusted later down the line. We need to take care of conflict resolution first.

Commit:
Merge branch '9.0' into publishingBonanza

Commit:
Adjust to Neos dev distribution adjustments

Commit:
Adjust to exposed rebase error handling strategy

mhsdesign commented on Classes/Controller/BackendServiceController.php:
```diff
@@ -182,56 +170,70 @@ public function changeAction(array $changes): void
-    public function publishAllAction(): void
+    public function publishChangesInSiteAction(array $command): void
     {
+        /** @todo send from UI */
```
todo remove todo, or later

mhsdesign reviewed (approved):
Thanks for the change i will run the Neos Ui e2e locally and then i suppose we can merge this.

grebaldi commented:
Just a reminder:

- https://github.com/neos/neos-development-collection/pull/4943 should be merged first
- https://github.com/neos/neos-ui/commit/3a06560087b6c7736e5175dcbc4b6d312c653e8c must be reverted before this one gets merged

mhsdesign commented:
Locally all tests pass, except an unrelated unit test failure that should be fixed after an upmerge. Ill push also a commit to fix the silly ci (sorry i meant style ci)

mhsdesign committed:
TASK: Fix style ci

mhsdesign committed:
TASK: Rename `WorkspaceFactory` to `WorkspaceProvider`

mhsdesign committed:
TASK: Rename to `retrieveForWorkspaceName`

Commit:
Adjust to new workspace provider methods

mhsdesign commented:
reopen to trigger ci

mhsdesign commented:
Huh e2e still fail

> Fatal error: Cannot redeclare Neos\Neos\Domain\Workspace\Workspace::getCurrentContentStreamId() in /home/circleci/app/Packages/Neos/Neos.Neos/Classes/Domain/Workspace/Workspace.php on line 131

also on 9.0 after merge ...

mhsdesign commented on Classes/Controller/BackendServiceController.php:
```diff
@@ -182,56 +170,70 @@ public function changeAction(array $changes): void
-            $nodeIdentifiersToPublish = [];
-            foreach ($nodeContextPaths as $contextPath) {
-                $nodeAddress = $nodeAddressFactory->createFromUriString($contextPath);
-                $nodeIdentifiersToPublish[] = new NodeIdToPublishOrDiscard(
-                    $nodeAddress->nodeAggregateId,
-                    $nodeAddress->dimensionSpacePoint
-                );
```
FYI, as we dont use the node address from the Neos Ui here anymore and their dimension we accidentally trashed partial publishing with respect to dimensions.

But this might be okay as discussed :D
